File: src/main/java/com/alphawallet/attestation/cheque/Cheque.java
package com.alphawallet.attestation.cheque;

import com.alphawallet.attestation.IdentifierAttestation.AttestationType;
import com.alphawallet.attestation.core.Attestable;
import com.alphawallet.attestation.core.AttestationCrypto;
import com.alphawallet.attestation.core.ExceptionUtil;
import com.alphawallet.attestation.core.SignatureUtility;
import java.io.IOException;
import java.math.BigInteger;
import java.time.Clock;
import java.util.Date;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bouncycastle.asn1.ASN1Encodable;
import org.bouncycastle.asn1.ASN1EncodableVector;
import org.bouncycastle.asn1.ASN1GeneralizedTime;
import org.bouncycastle.asn1.ASN1Integer;
import org.bouncycastle.asn1.ASN1Sequence;
import org.bouncycastle.asn1.DERBitString;
import org.bouncycastle.asn1.DEROctetString;
import org.bouncycastle.asn1.DERSequence;
import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
import org.bouncycastle.crypto.AsymmetricCipherKeyPair;
import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
import org.bouncycastle.crypto.util.SubjectPublicKeyInfoFactory;

public class Cheque implements Attestable {
  private static final Logger logger = LogManager.getLogger(Cheque.class);
  private final byte[] commitment;
  private final long amount;
  private final long notValidBefore;
  private final long notValidAfter;
  private final AsymmetricKeyParameter publicKey;
  private final byte[] signature;

  private final byte[] encoded;

  /**
   *
   * @param identifier The identifier of the receiver
   * @param type The type of identifier given
   * @param amount Amount of units the cheque should be valid for
   * @param validity time from now which the cheque should be valid, in milliseconds
   * @param keys the keys used to sign the cheque
   * @param secret the secret that must be known to cash the cheque
   */
  public Cheque(String identifier, AttestationType type, long amount, long validity, AsymmetricCipherKeyPair keys, BigInteger secret) {
    this.commitment = AttestationCrypto.makeCommitment(identifier, type, secret);
    this.publicKey = keys.getPublic();
    this.amount = amount;
    long current =  Clock.systemUTC().millis();
    this.notValidBefore = current - (current % 1000); // Round down to nearest second
    this.notValidAfter = this.notValidBefore + validity;
    ASN1Sequence cheque = makeCheque(this.commitment, amount, notValidBefore, notValidAfter);
    try {
      this.signature = SignatureUtility.signWithEthereum(cheque.getEncoded(), keys.getPrivate());
      this.encoded = encodeSignedCheque(cheque, this.signature, this.publicKey);
    } catch (IOException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not encode asn1", e);
    }
    if (!verify()) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("Could not verify object"));
    }
  }

  public Cheque(byte[] commitment, long amount, long notValidBefore, long notValidAfter, byte[] signature, AsymmetricKeyParameter publicKey) {
    this.commitment = commitment;
    this.publicKey = publicKey;
    this.amount = amount;
    if (notValidBefore % 1000 != 0 || notValidAfter % 1000 != 0) {
      throw ExceptionUtil.throwException(logger,
          new IllegalArgumentException("Can only support time granularity to the second"));
    }
    this.notValidBefore = notValidBefore;
    this.notValidAfter = notValidAfter;
    this.signature = signature;
    ASN1Sequence cheque = makeCheque(this.commitment, amount, notValidBefore, notValidAfter);
    try {
      this.encoded = encodeSignedCheque(cheque, this.signature, this.publicKey);
    } catch (IOException e) {
      throw ExceptionUtil.makeRuntimeException(logger, "Could not encode asn1", e);
    }
    if (!verify()) {
      throw ExceptionUtil.throwException(logger, new IllegalArgumentException("Signature is invalid"));
    }
  }

  private ASN1Sequence makeCheque(byte[] commitment, long amount, long notValidBefore, long notValidAfter) {
    ASN1EncodableVector cheque = new ASN1EncodableVector();
    cheque.add(new ASN1Integer(amount));

    ASN1GeneralizedTime notValidBeforeEnc = new ASN1GeneralizedTime(new Date(notValidBefore));
    ASN1GeneralizedTime notValidAfterEnc = new ASN1GeneralizedTime(new Date(notValidAfter));
    ASN1Sequence validityEnc = new DERSequence(new ASN1Encodable[] {notValidBeforeEnc, notValidAfterEnc});
    cheque.add(validityEnc);

    cheque.add(new DEROctetString(commitment));

    return new DERSequence(cheque);
  }

  private byte[] encodeSignedCheque(ASN1Sequence cheque, byte[] signature, AsymmetricKeyParameter publicKey) throws IOException {
      ASN1EncodableVector signedCheque = new ASN1EncodableVector();
      signedCheque.add(cheque);

      SubjectPublicKeyInfo spki = SubjectPublicKeyInfoFactory.createSubjectPublicKeyInfo(publicKey);
      signedCheque.add(spki.getPublicKeyData());

      signedCheque.add(new DERBitString(signature));
      return new DERSequence(signedCheque).getEncoded();
  }

  @Override
  public boolean checkValidity() {
    long currentTime = Clock.systemUTC().millis();
    if (!(currentTime >= getNotValidBefore() && currentTime < getNotValidAfter())) {
      logger.error("Cheque is no longer valid");
      return false;
    }
    return true;
  }

  @Override
  public boolean verify() {
    try {
      ASN1Sequence cheque = makeCheque(this.commitment, this.amount, this.getNotValidBefore(),
          this.notValidAfter);
      if (!SignatureUtility.verifyEthereumSignature(cheque.getEncoded(), signature, this.publicKey)) {
        logger.error("Could not verify signature");
        return false;
      }
    } catch (IOException e) {
      logger.error("Could not decode signature");
      return false;
    }
    return true;
  }

  @Override
  public byte[] getDerEncoding() {
    return encoded;
  }

  @Override
  public byte[] getCommitment() {
    return commitment;
  }

  public long getAmount() {
    return amount;
  }

  public long getNotValidBefore() {
    return notValidBefore;
  }

  public long getNotValidAfter() {
    return notValidAfter;
  }

  public byte[] getSignature() {
    return signature;
  }

  public AsymmetricKeyParameter getPublicKey() {
    return this.publicKey;
  }
}

File: src/main/javascript/crypto/src/libs/Cheque.ts
import {ATTESTATION_TYPE} from "./interfaces";
import {Asn1Der} from "./DerUtility";
import { AttestationCrypto } from "./AttestationCrypto";
import {hexStringToArray, uint8tohex} from "./utils";
import {KeyPair} from "./KeyPair";
import {SignatureUtility} from "./SignatureUtility";

let sha3 = require("js-sha3");
let EC = require("elliptic");
let ec = new EC.ec('secp256k1');


export class Cheque {
    // publicKey: string;
    // riddle: Uint8Array;
    private commitment: Uint8Array;
    public encoded: string;
    private identifier: string;
    private type: string;
    private amount: number;
    private validity: number;
    private keys: KeyPair;
    private secret: bigint;
    private notValidBefore: number;
    private notValidAfter: number;
    private signature: Uint8Array;
    // TODO code it
    constructor() {}

    static fromData(commitment: Uint8Array, amount: number, notValidBefore: number, notValidAfter: number, signature: Uint8Array, keys: KeyPair) {
        const me = new this();

        me.commitment = commitment;
        me.keys = keys;
        me.amount = amount;
        if (notValidBefore % 1000 != 0 || notValidAfter % 1000 != 0) {
            throw new Error("Can only support time granularity to the second");
        }
        me.notValidBefore = notValidBefore;
        me.notValidAfter = notValidAfter;
        me.signature = signature;
        let cheque = me.makeCheque();
        me.encoded = me.encodeSignedCheque(cheque, uint8tohex(signature));

        if (!me.verify()) {
            throw new Error("Signature is invalid");
        }

        return me;
    }

    static createAndVerify(identifier: string, type: string, amount: number, validity: number, keys: KeyPair, secret: bigint){

        let me = new this();

        me.identifier = identifier;
        me.type = type;
        me.amount = amount;
        me.validity = validity;
        me.keys = keys;
        me.secret = secret;

        let crypto = new AttestationCrypto();
        me.commitment = crypto.makeCommitment(me.identifier, ATTESTATION_TYPE[me.type], me.secret);

        let current =  new Date().getTime() ;
        me.notValidBefore = current - (current % 1000); // Round down to nearest second
        me.notValidAfter = me.notValidBefore + me.validity * 1000;
        let cheque = me.makeCheque();

        let ecKey = ec.keyFromPrivate(me.keys.getPrivateAsHexString(), 'hex');
        let chequeHash = sha3.keccak256(hexStringToArray(cheque));
        var signature = ecKey.sign( chequeHash );

        let signatureHexDerDitString = Asn1Der.encode('BIT_STRING', signature.toDER('hex'));

        me.encoded = me.encodeSignedCheque(
            cheque,
            signatureHexDerDitString,
        );

        let verify = ecKey.verify(chequeHash, signature);

        if (!verify) {
            throw new Error("Public and private keys are incorrect");
        }
        return {
            cheque,
            chequeEncoded: me.encoded,
            derSignature: signatureHexDerDitString,
            derSecret: Asn1Der.encode('SEQUENCE_30', Asn1Der.encode('OCTET_STRING', me.secret.toString(16)))
        }
    }

    encodeSignedCheque(cheque: string, signature: string){
        let fullSequence = cheque + Asn1Der.encode('BIT_STRING', this.keys.getPublicKeyAsHexStr()) + signature;
        return Asn1Der.encode('SEQUENCE_30', fullSequence);
    }

    makeCheque(){
        let timeList =
            Asn1Der.encode('GENERALIZED_TIME', formatGeneralizedDateTime(this.notValidBefore)) +
            Asn1Der.encode('GENERALIZED_TIME', formatGeneralizedDateTime(this.notValidAfter));
        let fullSequence =
            Asn1Der.encode('INTEGER', this.amount) +
            Asn1Der.encode('SEQUENCE_30', timeList) +
            Asn1Der.encode('OCTET_STRING', uint8tohex(this.commitment));
        return Asn1Der.encode('SEQUENCE_30', fullSequence);
    }


    verify(): boolean{
        let cheque = this.makeCheque();
        return SignatureUtility.verify(cheque, uint8tohex(this.signature), this.keys);
    }

    // TODO code it
    getDerEncoding(): Uint8Array{
        return Uint8Array.from([]);
    }
    public checkValidity(): boolean {
        let now: number = Date.now();
        if ( this.notValidBefore > now ) {
            console.log("Cheque is no longer valid");
            return false;
        }

        if ( this.notValidAfter < now ) {
            console.log("Cheque expired");
            return false;
        }

        return true;
    }

    getCommitment(){
        return this.commitment;
    }

}

// TODO add timezone
function formatGeneralizedDateTime(date: any):string {
    var d = new Date(date),
        month = '' + (d.getUTCMonth() + 1),
        day = '' + d.getUTCDate(),
        year = d.getUTCFullYear();
    let hour = '' + d.getUTCHours(),
        min = '' + d.getUTCMinutes(),
        sec = '' + d.getUTCSeconds()

    if (month.length < 2)
        month = '0' + month;
    if (day.length < 2)
        day = '0' + day;
    if (hour.length < 2)
        hour = '0' + hour;
    if (min.length < 2)
        min = '0' + min;
    if (sec.length < 2)
        sec = '0' + sec;

    return [year, month, day, hour, min, sec].join('') + 'Z';
}
